libshpool/src/config_watcher.rs

```diff
@@ -0,0 +1,676 @@
+    notify_rx: Receiver<Result<notify::Event, notify::Error>>,
+    // receiving watch requests
+    req_rx: Receiver<PathBuf>,
+    // from target_path to (watched_path, immediate_child_path)
+    paths: HashMap<PathBuf, (PathBuf, PathBuf)>,
+
+    // for sending out debug info
+    #[cfg(test)]
+    debug_tx: Sender<()>,
+}
+
+// Outcomes of selecting channels in the worker thread
+enum Outcome {
+    // A notify event occurred
+    Event(notify::Result<notify::Event>),
+    // A control command from outside
+    AddWatch(PathBuf),
+    // Timeout on notify event, trigger reload
+    Timeout,
+    // Any channel was disconnected
+    Disconnected,
+}
+
+impl From<Result<notify::Result<notify::Event>, RecvError>> for Outcome {
+    fn from(value: Result<notify::Result<notify::Event>, RecvError>) -> Self {
+        match value {
+            Ok(v) => Self::Event(v),
+            Err(RecvError) => Self::Disconnected,
+        }
+    }
+}
+
+impl From<Result<PathBuf, RecvError>> for Outcome {
+    fn from(value: Result<PathBuf, RecvError>) -> Self {
+        match value {
+            Ok(v) => Self::AddWatch(v),
+            Err(RecvError) => Self::Disconnected,
+        }
+    }
+}
+
+impl<Handler> ConfigWatcherInner<Handler> {
+    // get next event to work on
+    fn select(&self) -> Outcome {
+        debug!("now {:?} select with ddl {:?}", Instant::now(), &self.reload_deadline);
+
+        // only impose a deadline if there is pending reload
+        let timeout = self
+            .reload_deadline
+            .map(crossbeam_channel::at)
+            .unwrap_or_else(crossbeam_channel::never);
+        // first try poll
+        if let Ok(res) = self.notify_rx.try_recv() {
+            return Outcome::Event(res);
+        }
+        if let Ok(res) = self.req_rx.try_recv() {
+            return Outcome::AddWatch(res);
+        }
+        if timeout.try_recv().is_ok() {
+            return Outcome::Timeout;
+        }
+
+        // nothing ready to act immediately, notify debug_tx
+        #[cfg(test)]
+        self.debug_tx.send(()).unwrap();
+
+        // finally blocking wait
+        let res = select! {
+            recv(self.notify_rx) -> res => Outcome::from(res),
+            recv(self.req_rx) -> res => Outcome::from(res),
+            recv(timeout) -> _ => Outcome::Timeout,
+        };
+        res
+    }
+
+    fn trigger_reload(&mut self) {
+        self.reload_deadline =
+            self.reload_deadline.or_else(|| Some(Instant::now() + self.reload_debounce));
+        debug!("defer config reloading to {:?}!", &self.reload_deadline);
+    }
+}
+
+impl<Handler> ConfigWatcherInner<Handler>
+where
+    Handler: FnMut(),
+{
+    // run forever loop to reload config, only return when there is error to create any watches.
+    fn run(&mut self) -> anyhow::Result<()> {
+        loop {
+            match self.select() {
+                Outcome::Event(res) => {
+                    let (rewatch, mut reload) = match res {
```

Comment:
why do we need to rewatch paths? (just curious, I'm sure there is a good reason for it)

Reply:
The idea is that if any segment in the target path is created/removed, we'll need to update our watched path to watch the next/previous segment. This is needed because we use non-recursive watch, so only we only get events to the watched path and its immediate children.

We can't simply watch recursively on the whole fs tree, that would add too many watches in the system and can easily exhaust available watchers.

Rewatch is basically re-walk the target_path to find the new cloest parent to add a watch.

Probably easier to explain using an example:

* target_path: `/base/sub/config.toml`
* watched_path: `/base`
* immediate_child: `sub`

a) if watched_path (`/base`) is removed, we need to rewatch on `/` => new watched_path is `/`, new immediate_child is `base`
b) if immediate_child (`sub`) is created, we need to rewatch on `/base/sub` => new watched_path is `/base/sub`, new immediate_child is `config.toml`
c) other combinations (watched_path created, immediate_child removed) are impossible in theory, but might happen if some events are missed. This should be covered by [`need_rescan`](https://docs.rs/notify/latest/notify/event/enum.Flag.html#variant.Rescan), but we handle it the same way anyways as the logic is simpler.

Reply:
Oh yeah that makes lots of sense.